core/state_processor.go

```diff
@@ -98,23 +101,50 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 		allLogs = append(allLogs, receipt.Logs...)
 	}
 
+	// Overlay tree migration logic
+	migrdb := statedb.Database()
+	// TODO: avoid opening the preimages file here and make it part of, potentially, statedb.Database().
+	filePreimages, err := os.Open("preimages.bin")
+	if err != nil {
+		return nil, nil, 0, fmt.Errorf("opening preimage file: %s", err)
+	}
+	defer filePreimages.Close()
```

Comment:
As we discussed, left this here to avoid rebase conflicts. We can figure out in the future the right place for this to live.